File: examples/java/customer/src/main/java/dev/angzarr/examples/customer/CustomerService.java
package dev.angzarr.examples.customer;

import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;
import examples.Domains.AddLoyaltyPoints;
import examples.Domains.CreateCustomer;
import examples.Domains.RedeemLoyaltyPoints;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import dev.angzarr.BusinessLogicGrpc;
import dev.angzarr.BusinessResponse;
import dev.angzarr.CommandBook;
import dev.angzarr.ContextualCommand;
import dev.angzarr.EventBook;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * gRPC service implementation for customer business logic.
 * Uses dependency injection for the business logic implementation.
 */
public class CustomerService extends BusinessLogicGrpc.BusinessLogicImplBase {
    private static final Logger logger = LoggerFactory.getLogger(CustomerService.class);
    private static final String DOMAIN = "customer";

    private final CustomerLogic logic;

    public CustomerService(CustomerLogic logic) {
        this.logic = logic;
    }

    @Override
    public void handle(ContextualCommand request, StreamObserver<BusinessResponse> responseObserver) {
        try {
            EventBook events = processCommand(request);
            BusinessResponse response = BusinessResponse.newBuilder()
                .setEvents(events)
                .build();
            responseObserver.onNext(response);
            responseObserver.onCompleted();
        } catch (CommandValidationException e) {
            responseObserver.onError(Status.fromCode(e.getStatusCode())
                .withDescription(e.getMessage())
                .asRuntimeException());
        } catch (InvalidProtocolBufferException e) {
            responseObserver.onError(Status.INVALID_ARGUMENT
                .withDescription("Failed to parse command: " + e.getMessage())
                .asRuntimeException());
        } catch (Exception e) {
            logger.error("Unexpected error processing command", e);
            responseObserver.onError(Status.INTERNAL
                .withDescription("Internal error: " + e.getMessage())
                .asRuntimeException());
        }
    }

    private EventBook processCommand(ContextualCommand request)
            throws CommandValidationException, InvalidProtocolBufferException {
        CommandBook cmdBook = request.getCommand();
        EventBook priorEvents = request.getEvents();

        if (cmdBook == null || cmdBook.getPagesList().isEmpty()) {
            throw CommandValidationException.invalidArgument("CommandBook has no pages");
        }

        var cmdPage = cmdBook.getPages(0);
        if (!cmdPage.hasCommand()) {
            throw CommandValidationException.invalidArgument("Command page has no command");
        }

        CustomerState state = logic.rebuildState(priorEvents);
        Any command = cmdPage.getCommand();
        String typeUrl = command.getTypeUrl();

        EventBook result;
        if (typeUrl.endsWith("CreateCustomer")) {
            CreateCustomer cmd = command.unpack(CreateCustomer.class);
            result = logic.handleCreateCustomer(state, cmd.getName(), cmd.getEmail());
        } else if (typeUrl.endsWith("AddLoyaltyPoints")) {
            AddLoyaltyPoints cmd = command.unpack(AddLoyaltyPoints.class);
            result = logic.handleAddLoyaltyPoints(state, cmd.getPoints(), cmd.getReason());
        } else if (typeUrl.endsWith("RedeemLoyaltyPoints")) {
            RedeemLoyaltyPoints cmd = command.unpack(RedeemLoyaltyPoints.class);
            result = logic.handleRedeemLoyaltyPoints(state, cmd.getPoints(), cmd.getRedemptionType());
        } else {
            throw CommandValidationException.invalidArgument("Unknown command type: " + typeUrl);
        }

        // Add cover from command book to the result
        return result.toBuilder()
            .setCover(cmdBook.getCover())
            .build();
    }
}

File: examples/kotlin/customer/src/main/kotlin/dev/angzarr/examples/customer/CustomerService.kt
package dev.angzarr.examples.customer

import dev.angzarr.BusinessLogicGrpcKt
import dev.angzarr.BusinessResponse
import dev.angzarr.ContextualCommand
import examples.Domains.AddLoyaltyPoints
import examples.Domains.CreateCustomer
import examples.Domains.RedeemLoyaltyPoints
import io.grpc.Status

/**
 * gRPC service adapter for customer business logic.
 * Thin wrapper that delegates to CustomerLogic and maps exceptions to gRPC status codes.
 */
class CustomerService(
    private val logic: CustomerLogic
) : BusinessLogicGrpcKt.BusinessLogicCoroutineImplBase() {

    override suspend fun handle(request: ContextualCommand): BusinessResponse {
        val cmdBook = request.command
        val priorEvents = request.events

        if (cmdBook.pagesList.isEmpty()) {
            throw Status.INVALID_ARGUMENT
                .withDescription("CommandBook has no pages")
                .asRuntimeException()
        }

        val cmdPage = cmdBook.pagesList[0]
        val cmd = cmdPage.command ?: throw Status.INVALID_ARGUMENT
            .withDescription("Command page has no command")
            .asRuntimeException()

        val state = logic.rebuildState(priorEvents)

        try {
            val eventBook = when {
                cmd.`is`(CreateCustomer::class.java) -> {
                    val c = cmd.unpack(CreateCustomer::class.java)
                    logic.handleCreateCustomer(state, c.name, c.email)
                }
                cmd.`is`(AddLoyaltyPoints::class.java) -> {
                    val c = cmd.unpack(AddLoyaltyPoints::class.java)
                    logic.handleAddLoyaltyPoints(state, c.points, c.reason)
                }
                cmd.`is`(RedeemLoyaltyPoints::class.java) -> {
                    val c = cmd.unpack(RedeemLoyaltyPoints::class.java)
                    logic.handleRedeemLoyaltyPoints(state, c.points, c.redemptionType)
                }
                else -> throw Status.INVALID_ARGUMENT
                    .withDescription("Unknown command type: ${cmd.typeUrl}")
                    .asRuntimeException()
            }

            // Copy cover from command book to event book
            val eventBookWithCover = eventBook.toBuilder()
                .setCover(cmdBook.cover)
                .build()

            return BusinessResponse.newBuilder()
                .setEvents(eventBookWithCover)
                .build()

        } catch (e: CommandValidationException) {
            throw Status.fromCode(e.statusCode)
                .withDescription(e.message)
                .asRuntimeException()
        }
    }
}
